### PR TITLE
(PT-789) Add documentation for config policies and `policy` subcommands

### PR DESCRIPTION
**Pull request by rileynewton**
Added documentation on policies and a guide on what has been added to Bolt vis-à-vis policies.

### CODE REVIEW

File: documentation/experimental_features.md
Comment (lucywyman) [inline documentation/experimental_features.md:16]:
```suggestion
or more nodes. Policies are Puppet code stored in the `manifests/` directory of modules on the modulepath.
```

File: documentation/experimental_features.md
Comment (lucywyman) [inline documentation/experimental_features.md:17]:
```suggestion
```

File: documentation/experimental_features.md
Comment (lucywyman) [inline documentation/experimental_features.md:15]:
```suggestion
Configuration policies, or policies for short, are public classes that can be applied directly to one
```

Comment (lucywyman) [reply documentation/experimental_features.md:15]: Assigning policies to nodes in a way that persists beyond one application requires what we call a "classifier" or external node classifier, which is only going to exist in Cygnus, not Bolt.

File: documentation/experimental_features.md
Comment (lucywyman) [inline documentation/experimental_features.md:19]:
```suggestion
Policies can be managed through the `policy` subcommand, with `apply`, `new`
```

File: documentation/experimental_features.md
Comment (lucywyman) [inline documentation/experimental_features.md:20]:
```suggestion
and `show` subcommands. The default location for a policy created through Bolt is
```

File: documentation/experimental_features.md
Comment (lucywyman) [inline documentation/experimental_features.md:15]: I think it's important to include in this top paragraph that policies are only different from classes in that they are listed in `bolt-project.yaml` under the `policies` key. Something like:
```
Configuration policies are public classes that can be assigned directly to one or more nodes via the Bolt command line. A class becomes a policy when it's listed under the `policies` key in `bolt-project.yaml`, and if the `policies` key is empty then no policies will be available for Bolt to apply.
```

File: documentation/experimental_features.md
Comment (lucywyman) [inline documentation/experimental_features.md:24]:
```suggestion
You can list available policies using `bolt policy show`, and apply them to targets using `bolt policy apply`.
```

**Review by beechtom — COMMENTED** [review]
Overall I think this is a good start, but it would be helpful to provide users with a bit more context and give this section a bit more structure. Since this is a completely new feature and concept that isn't documented anywhere else, I think it's important that we err on the side of giving users more information so they understand what policies are and how to use these commands. An outline might look like:

- Summary of what policies are
  - Classes on the modulepath in a module's or project's `manifests/` directory
  - Can be applied directly to targets just like other Puppet code
- Listing available policies
  - Policies must be listed under the `policies` key of the project config file to be available to the project
  - Supports glob patterns for easily listing multiple policies (i.e. `policies::*` includes the `policies::admin` and `policies::sshkeys` policies)
  - Example project config file and the shell/powershell commands to list policies, with sample output
- Creating new policies
  - Policies can be created by manually adding it to a module's or project's `manifests/` directory
  - You can also use a Bolt command to create the policy for you
  - Policy names must follow [class naming conventions](https://puppet.com/docs/puppet/7/lang_reserved.html#classes-and-defined-resource-type-names)
  - When Bolt creates a policy it adds it to the project's `manifests/` directory and adds the policy to the list of available policies in the project config
  - Example commands with output
- Applying policies to targets
  - Similar to applying Puppet code with `bolt apply`
  - Allows you to apply one or more policies at a time
  - Before applying policies, will install the puppet-agent package and collect facts for each target
  - Example policy, commands that apply the policy to targets, and sample output

File: documentation/experimental_features.md
Comment (lucywyman) [inline documentation/experimental_features.md:28]: "Bolt will create a single line of Puppet code `include #{policies.join(,)}` that will be compiled and applied to the provided targets, and will log that code at the debug level."
I think this will clarify some questions that people who are more experienced may have around how classes are applied, are they applied in order, etc.

File: documentation/experimental_features.md
Comment (lucywyman) [inline documentation/experimental_features.md:32]:
```suggestion
project. When not configured, no policies will be available to Bolt to be applied.
```
I mostly just don't want people to think they can't use `bolt policy new` in this case, since they can.

File: documentation/experimental_features.md
Comment (lucywyman) [inline documentation/experimental_features.md:11]:
```suggestion
## Configuration policies and the policy command
```

File: documentation/experimental_features.md
Comment (rileynewton) [reply documentation/experimental_features.md:19]: Great, yeah I was being terse here.

File: documentation/experimental_features.md
Comment (beechtom) [inline documentation/experimental_features.md:63]:
```suggestion
2.  An empty class is created in the project's `manifests/` directory that you
    can then populate. Under the hood, Bolt creates a single
    line of Puppet code `include <POLICIES>` that will be compiled and
    applied to the provided targets, and will log that code at the debug level.
```

- Bolt will create a nested directory structure if the policy's name has more than two name segments, so that example path isn't accurate. Since Bolt displays the path to the file it creates, I don't think there's a need to document it here.
- We should avoid using Ruby code to tell users what is happening under the hood. I think the bit about what's happening under the hood should also move to the `policy apply` section.

File: documentation/experimental_features.md
Comment (beechtom) [inline documentation/experimental_features.md:64]:
```suggestion
For example, to create a new project-level policy named `bolt::project::user`, run:
```

Each of the examples should have a quick blurb about what it's doing.

File: documentation/experimental_features.md
Comment (beechtom) [inline documentation/experimental_features.md]:
```suggestion
Policy names must follow [class naming
conventions](https://puppet.com/docs/puppet/7/lang_reserved.html#classes-and-defined-resource-type-names).
```

Policy names can have more than two name segments, so leave out the bit about it having to be in the form `<PROJECT NAME>::<POLICY NAME>`. If we want, we can add examples of valid and invalid policy names.

We should also move this bit down so it's under the example commands.

File: documentation/experimental_features.md
Comment (rileynewton) [reply documentation/experimental_features.md]: Made this eliminate redundancy change before review, changed language to 'not
configured' as suggested.

File: documentation/experimental_features.md
Comment (rileynewton) [reply documentation/experimental_features.md]: I agree, not needed with the better structuring of sections, removed.

File: documentation/experimental_features.md
Comment (rileynewton) [reply documentation/experimental_features.md]: Emulated that structure.

File: documentation/experimental_features.md
Comment (rileynewton) [reply documentation/experimental_features.md]: Made those changes, changed examples to using a nested structure as well which we may not want.

File: documentation/experimental_features.md
Comment (beechtom) [inline documentation/experimental_features.md:57]:
```suggestion
  New-BoltPolicy -Name bolt::project::user
```

File: documentation/experimental_features.md
Comment (rileynewton) [reply documentation/experimental_features.md]: Changed.

File: documentation/experimental_features.md
Comment (beechtom) [inline documentation/experimental_features.md:32]:
```suggestion
multiple policies. For example, `bolt::project::*` includes `bolt::project::admin` and
`bolt::project::sshkeys` policies.
```

We should also make sure the doc is consistent between `boltproject` and `bolt::project` prefixes.

File: documentation/experimental_features.md
Comment (rileynewton) [reply documentation/experimental_features.md:62]: Covered the blurbs I think.

File: documentation/experimental_features.md
Comment (beechtom) [inline documentation/experimental_features.md:119]:
```suggestion
  Invoke-BoltPolicy -Name bolt::project::admin,bolt::project::sshkeys -Targets mytarget
```

File: documentation/experimental_features.md
Comment (beechtom) [inline documentation/experimental_features.md:78]:
```suggestion
You can list available policies for the project:
```

File: documentation/experimental_features.md
Comment (rileynewton) [reply documentation/experimental_features.md]: Removed the name segments, put the policy name around the example commands, though its placing still seems awkward.

File: documentation/experimental_features.md
Comment (beechtom) [inline documentation/experimental_features.md:68]:
```suggestion
Policy names must follow [class naming
conventions](https://puppet.com/docs/puppet/7/lang_reserved.html#classes-and-defined-resource-type-names).
```

Hm, I think this might work better as the first line at line 60.

**Review by beechtom — COMMENTED** [review]
This is looking good 👍  Just a few last nits.

File: documentation/experimental_features.md
Comment (rileynewton) [reply documentation/experimental_features.md:57]: Was going to ask about getting the Powershell scripts in case there was an added param like this. :)

File: documentation/experimental_features.md
Comment (rileynewton) [reply documentation/experimental_features.md:32]: Missed this, thanks.

**Review by beechtom — APPROVED** [review]

File: documentation/experimental_features.md
Comment (hestonhoffman) [inline documentation/experimental_features.md:38]:
```suggestion
# bolt-project.yaml
name: myproject
```
Just reiterating for the user where the config file is. Also, I think using `bolt` as the project name might be a tad confusing, so changing that to something more generic

File: documentation/experimental_features.md
Comment (hestonhoffman) [inline documentation/experimental_features.md:63]:
```suggestion
Bolt creates an empty class in the project's `manifests/` directory that you can
populate with code. Policy names must follow [class naming
conventions](https://puppet.com/docs/puppet/7/lang_reserved.html#classes-and-defined-resource-type-names).

```

File: documentation/experimental_features.md
Comment (hestonhoffman) [inline documentation/experimental_features.md:70]:
```suggestion
Create policies manually by:
```

File: documentation/experimental_features.md
Comment (hestonhoffman) [inline documentation/experimental_features.md:107]:
```suggestion
Applying policies is similar to [applying Puppet
code](https://puppet.com/docs/bolt/latest/applying_manifest_blocks.html), 
with the addition that you can apply one or more policies at a time. The Bolt 
commands for applying policies accept a single policy name or a 
comma-separated list of policy names to apply to a list of one or more targets. 
```

File: documentation/experimental_features.md
Comment (hestonhoffman) [inline documentation/experimental_features.md:138]:
```suggestion
Before applying policies, Bolt installs the `puppet-agent` package and collects
facts from each target. Bolt creates a single line of Puppet code to compile 
and apply to the provided targets. This line of code includes the syntax
`include <POLICIES>` and Bolt logs the line at the debug level.
```

File: documentation/experimental_features.md
Comment (rileynewton) [reply documentation/experimental_features.md:38]: I thought it was a potentially confusing project name too, but @beechtom is this common Bolt convention to have the name be 'bolt'? With the current code, policy names always have to start with the project name, so I'd change the other references in the doc to "myproject::bolt::admin" or some other name in the middle.

Comment (beechtom) [reply documentation/experimental_features.md:37]: There's really no convention for project names. I personally name mine `bolt`, but I think the rest of our docs use `myproject` typically so I think it makes sense to do the same here.

There's also no need to have three name segments, so we could potentially simplify the policy names to `myproject::admin`, etc.

File: documentation/experimental_features.md
Comment (donoghuc) [inline documentation/experimental_features.md:134]: Would it be work linking to documentation about the nuances of apply_prep here (just blanket saying it installs the agent seems somewhat inaccurate and scary) Maybe if we dont want to find a link just say something like `Bolt ensures the puppet-agent packages is available on the target and collects facts`?

Comment (hestonhoffman) [reply documentation/experimental_features.md:134]: How does this sound?
> Before applying policies, Bolt installs uses the `apply_prep` function to collect 
  facts and ensure that the `puppet-agent` package is available on each target.
  For more information on how Bolt uses `apply_prep`, see
  [Applying manifest blocks]applying_manifest_blocks.md#applying-manifest-blocks-from-a-puppet-plan

Comment (rileynewton) [reply documentation/experimental_features.md:134]: That sounds good to me, does this check out with you @beechtom? I believe we can go ahead and merge this after this change unless @lucywyman wants to add anything.

Comment (beechtom) [reply documentation/experimental_features.md:134]: @rileynewton Looks good to me 👍 Before merging you'll need to squash all your commits down to a single commit.

File: documentation/experimental_features.md
Comment (donoghuc) [inline documentation/experimental_features.md:136]:
```suggestion
Before applying policies, Bolt uses the `apply_prep` function to collect facts and
ensure that the puppet-agent package is available on each target. For more
information on how Bolt uses `apply_prep`, see [Applying manifest
```
Do we generally use unformatted text when referencing functions?

**Review by donoghuc — APPROVED** [review]
Looks good. Dont forget to squash the commits before merging ;)

File: documentation/experimental_features.md
Comment (rileynewton) [reply documentation/experimental_features.md:136]: Doh, committing.

**Pull request merged**
